File: experiments/engine-kt/adapter-openai-responses/src/main/kotlin/com/garive/eng/kt/openai/Sse.kt
package com.garive.eng.kt.openai

/** Incremental UTF-8 Server-Sent Events frame. */
public data class SseFrame(
    public val event: String?,
    public val data: String,
    public val id: String?,
    public val retry: ULong?,
)

/** Byte-chunk incremental SSE decoder. */
public class SseDecoder {
    private var buffer: ByteArray = byteArrayOf()

    /** Appends arbitrary transport bytes and emits complete frames immediately. */
    public fun push(bytes: ByteArray): List<SseFrame> = responseFailure(ResponsesProtocolError.INVALID_SSE) {
        buffer += bytes
        val frames = mutableListOf<SseFrame>()
        while (true) {
            val boundary = findBoundary(buffer) ?: break
            val frame = parseFrame(buffer.copyOfRange(0, boundary.first))
            buffer = buffer.copyOfRange(boundary.first + boundary.second, buffer.size)
            if (frame != null) frames += frame
        }
        return frames
    }

    /** Requires EOF to follow a complete frame or comments only. */
    public fun finish(): Unit = responseFailure(ResponsesProtocolError.TRUNCATED_STREAM) {
        val trailing = runCatching {
            buffer.decodeToString(throwOnInvalidSequence = true)
        }.getOrNull()
        require(trailing != null && trailing.lineSequence().all { it.isBlank() || it.startsWith(':') })
        buffer = byteArrayOf()
    }
}

private fun findBoundary(bytes: ByteArray): Pair<Int, Int>? {
    for (index in bytes.indices) {
        if (index + 1 < bytes.size && bytes[index] == 10.toByte() && bytes[index + 1] == 10.toByte()) {
            return index to 2
        }
        if (index + 3 < bytes.size && bytes[index] == 13.toByte() &&
            bytes[index + 1] == 10.toByte() && bytes[index + 2] == 13.toByte() &&
            bytes[index + 3] == 10.toByte()
        ) {
            return index to 4
        }
    }
    return null
}

private fun parseFrame(bytes: ByteArray): SseFrame? {
    val text = bytes.decodeToString(throwOnInvalidSequence = true)
    var event: String? = null
    var id: String? = null
    var retry: ULong? = null
    val data = mutableListOf<String>()
    text.lineSequence().forEach { raw ->
        val line = raw.removeSuffix("\r")
        if (line.isNotEmpty() && !line.startsWith(':')) {
            val field = line.substringBefore(':')
            val value = line.substringAfter(':', "").removePrefix(" ")
            when (field) {
                "event" -> event = value
                "data" -> data += value
                "id" -> if ('\u0000' !in value) id = value
                "retry" -> retry = value.toULong()
            }
        }
    }
    return if (data.isEmpty()) null else SseFrame(event, data.joinToString("\n"), id, retry)
}

File: adapters/anthropic-messages/src/sse.rs
//! Incremental UTF-8 Server-Sent Events framing independent of JSON semantics.

use crate::MessagesAdapterError;

/// One complete Server-Sent Events frame.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SseFrame {
    event: Option<String>,
    data: String,
    id: Option<String>,
    retry: Option<u64>,
}

impl SseFrame {
    /// Returns the optional SSE event name.
    pub fn event(&self) -> Option<&str> {
        self.event.as_deref()
    }
    /// Returns data lines joined with a single newline.
    pub fn data(&self) -> &str {
        &self.data
    }
    /// Returns the optional SSE event identifier.
    pub fn id(&self) -> Option<&str> {
        self.id.as_deref()
    }
    /// Returns the optional retry value in milliseconds.
    pub fn retry(&self) -> Option<u64> {
        self.retry
    }
}

/// Stateful SSE framer that accepts arbitrary transport byte chunks.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct SseDecoder {
    buffer: Vec<u8>,
}

impl SseDecoder {
    /// Creates an empty incremental decoder.
    pub fn new() -> Self {
        Self::default()
    }

    /// Appends bytes and returns every frame completed by this chunk.
    pub fn push(&mut self, bytes: &[u8]) -> Result<Vec<SseFrame>, MessagesAdapterError> {
        self.buffer.extend_from_slice(bytes);
        let mut frames = Vec::new();
        while let Some((end, delimiter_length)) = frame_boundary(&self.buffer) {
            let frame_bytes = self.buffer[..end].to_vec();
            self.buffer.drain(..end + delimiter_length);
            if let Some(frame) = parse_frame(&frame_bytes)? {
                frames.push(frame);
            }
        }
        Ok(frames)
    }

    /// Validates that EOF does not leave an incomplete frame.
    pub fn finish(&mut self) -> Result<(), MessagesAdapterError> {
        let trailing_comments = std::str::from_utf8(&self.buffer)
            .map(|text| {
                text.lines()
                    .all(|line| line.trim().is_empty() || line.starts_with(':'))
            })
            .unwrap_or(false);
        if trailing_comments {
            self.buffer.clear();
            Ok(())
        } else {
            Err(MessagesAdapterError::TruncatedStream)
        }
    }
}

fn frame_boundary(bytes: &[u8]) -> Option<(usize, usize)> {
    let lf = bytes.windows(2).position(|window| window == b"\n\n");
    let crlf = bytes.windows(4).position(|window| window == b"\r\n\r\n");
    match (lf, crlf) {
        (Some(left), Some(right)) if left <= right => Some((left, 2)),
        (Some(_), Some(right)) => Some((right, 4)),
        (Some(left), None) => Some((left, 2)),
        (None, Some(right)) => Some((right, 4)),
        (None, None) => None,
    }
}

fn parse_frame(bytes: &[u8]) -> Result<Option<SseFrame>, MessagesAdapterError> {
    let text = std::str::from_utf8(bytes).map_err(|_| MessagesAdapterError::InvalidSse)?;
    let mut event = None;
    let mut data = Vec::new();
    let mut id = None;
    let mut retry = None;
    for raw_line in text.split('\n') {
        let line = raw_line.strip_suffix('\r').unwrap_or(raw_line);
        if line.is_empty() || line.starts_with(':') {
            continue;
        }
        let (field, value) = line
            .split_once(':')
            .map(|(field, value)| (field, value.strip_prefix(' ').unwrap_or(value)))
            .unwrap_or((line, ""));
        match field {
            "event" => event = Some(value.to_owned()),
            "data" => data.push(value),
            "id" if !value.contains('\0') => id = Some(value.to_owned()),
            "retry" => {
                retry = Some(
                    value
                        .parse::<u64>()
                        .map_err(|_| MessagesAdapterError::InvalidSse)?,
                )
            }
            _ => {}
        }
    }
    if data.is_empty() {
        return Ok(None);
    }
    Ok(Some(SseFrame {
        event,
        data: data.join("\n"),
        id,
        retry,
    }))
}
